### PR TITLE
Optimize tooling around local dev environment

### PR DESCRIPTION
**Pull request by ccremer**
## Summary

This should make installing the tools required for local development easier.
The only thing we lose is having them versioned to a specific tag.
We've been using these tools for quite some time now and there are rarely issues than can be pointed to specific versions.
By running 'make clean', these tools get uninstalled, upon next time they're needed they get downloaded and updated again.

## Checklist

- [x] PR contains a single logical change (to keep release notes relevant)
- [x] Categorize the PR by setting a good title and adding one of the labels:
  `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
  as they show up in the changelog
- [x] I have run `make test-e2e` and e2e tests pass successfully

### CODE REVIEW

**Review by ZuGao — COMMENTED** [review]
Does it make sense to have so many targets for cleaning in so many files? I would go with just one target where you clean everything in one place.

**Comment by ccremer** [comment]
> Does it make sense to have so many targets for cleaning in so many files? I would go with just one target where you clean everything in one place.

I went for a more modular approach where it's more understandable why which files get to be cleaned.
If everything is cleaned in only 1 place, in 4 months you don't remember anymore from where a certain file or folder gets created, resp. you get scared to remove outdated files from this one place.
For example `.package-clean` cleans only crossplane-packaging related files. So if for some reason, we have no need for this packaging anymore, we can just remove `package.mk` and with it goes the cleaning as well. It goes the other way around too: If packaging gets more local-only files, then it can be easily added to `.package-clean` without worrying whether they are already part of the "global cleaning"

**Comment by Kidswiss** [comment]
@ZuGao I like the approach, this way you can clean individual tools, if you have issues with them.

File: kind/kind.mk [hunk 1]
```diff
@@ -1,13 +1,9 @@
...
+kind_bin = $(GOPATH)/bin/kind
 
 # Prepare kind binary
-# We need to set the Go arch since the binary is meant for the user's OS.
-$(kind_bin): export GOOS = $(shell go env GOOS)
-$(kind_bin): export GOARCH = $(shell go env GOARCH)
 $(kind_bin):
-	@mkdir -p $(kind_dir)
-	cd kind && go build -o $@ sigs.k8s.io/kind
+	go install sigs.k8s.io/kind@latest
```
Comment (Kidswiss) [inline kind/kind.mk:6]: This isn't working for me... for some reason I have to execute the `go install` outside the make:
```
gmake local-install
go install sigs.k8s.io/kind@latest
/Users/simonbeck/go/bin/kind create cluster \
        --name provider-cloudscale-v1.24.0 \
        --image docker.io/kindest/node:v1.24.0 \
        --config kind/config.yaml
/bin/bash: /Users/simonbeck/go/bin/kind: No such file or directory
gmake: *** [kind/kind.mk:36: /Users/simonbeck/repos/appcat/appcat-service-s3/.kind/kind-kubeconfig-v1.24.0] Error 127

go install sigs.k8s.io/kind@latest

gmake local-install               
/Users/simonbeck/go/bin/kind create cluster \
        --name provider-cloudscale-v1.24.0 \
        --image docker.io/kindest/node:v1.24.0 \
        --config kind/config.yaml
Creating cluster "provider-cloudscale-v1.24.0" ...
 ✓ Ensuring node image (docker.io/kindest/node:v1.24.0) 🖼 
 ✓ Preparing nodes 📦  
 ✓ Writing configuration 📜 
```

**Review by Kidswiss — CHANGES_REQUESTED** [review]
The installation of kind doesn't seem to work on my system from within the make file.

**Comment by Kidswiss** [comment]
I think I would actually prefer to keep the old tooling method.
1. it was versioned, now we're using latest which could introduce breaking changes at any time and break stuff at random
2. each project kept its own version, now we install them to the `$GOBIN`, which could overwrite existing versions

**Comment by ccremer** [comment]
> I think I would actually prefer to keep the old tooling method.
> 
>     1. it was versioned, now we're using latest which could introduce breaking changes at any time and break stuff at random
> 
>     2. each project kept its own version, now we install them to the `$GOBIN`, which could overwrite existing versions

I can understand why you want the old method. However, as pointed out in the PR description, since a year or so I've never experienced issues with tooling versions. By having them installed with `go install` we get rid of a lot of potential Renovate noise, and maybe more importantly: a LOT of boilerplate. `go install` supports specifying a specific version if a new version is not working for some reason.

We could consider setting `$GOBIN` in makefile to a local folder, e.g. `.work` instead of user's home. Would that be a better compromise regarding versioning issues?

File: kind/kind.mk (hunk 1 above)
Comment (ccremer) [reply kind/kind.mk:6]: Strange, I did not spot any make target dependency issues.
Can you reproduce it with debug flag? (`make -d`) so I can better understand what's happening.

**Comment by Kidswiss** [comment]
> We could consider setting $GOBIN in makefile to a local folder, e.g. .work instead of user's home. Would that be a better compromise regarding conflicting versions?

Yeah, I think this would be a good compromise.

File: kind/kind.mk (hunk 1 above)
Comment (Kidswiss) [reply kind/kind.mk:6]:
```
gmake -d local-install
GNU Make 4.3
Built for x86_64-apple-darwin21.1.0
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Reading makefile 'Makefile'...
Reading makefile 'Makefile.vars.mk' (search path) (no ~ expansion)...
Reading makefile 'docs/antora-preview.mk' (search path) (don't care) (no ~ expansion)...
Reading makefile 'docs/antora-build.mk' (search path) (don't care) (no ~ expansion)...
Reading makefile 'kind/kind.mk' (search path) (don't care) (no ~ expansion)...
Reading makefile 'test/local.mk' (search path) (don't care) (no ~ expansion)...
Reading makefile 'package/package.mk' (search path) (don't care) (no ~ expansion)...
Updating makefiles....
 Considering target file 'package/package.mk'.
  Looking for an implicit rule for 'package/package.mk'.
  No implicit rule found for 'package/package.mk'.
  Finished prerequisites of target file 'package/package.mk'.
 No need to remake target 'package/package.mk'.
 Considering target file 'test/local.mk'.
  Looking for an implicit rule for 'test/local.mk'.
  No implicit rule found for 'test/local.mk'.
  Finished prerequisites of target file 'test/local.mk'.
 No need to remake target 'test/local.mk'.
 Considering target file 'kind/kind.mk'.
  Looking for an implicit rule for 'kind/kind.mk'.
  No implicit rule found for 'kind/kind.mk'.
  Finished prerequisites of target file 'kind/kind.mk'.
 No need to remake target 'kind/kind.mk'.
 Considering target file 'docs/antora-build.mk'.
  Looking for an implicit rule for 'docs/antora-build.mk'.
  No implicit rule found for 'docs/antora-build.mk'.
  Finished prerequisites of target file 'docs/antora-build.mk'.
 No need to remake target 'docs/antora-build.mk'.
 Considering target file 'docs/antora-preview.mk'.
  Looking for an implicit rule for 'docs/antora-preview.mk'.
  No implicit rule found for 'docs/antora-preview.mk'.
  Finished prerequisites of target file 'docs/antora-preview.mk'.
 No need to remake target 'docs/antora-preview.mk'.
 Considering target file 'Makefile.vars.mk'.
  Looking for an implicit rule for 'Makefile.vars.mk'.
  No implicit rule found for 'Makefile.vars.mk'.
  Finished prerequisites of target file 'Makefile.vars.mk'.
 No need to remake target 'Makefile.vars.mk'.
 Considering target file 'Makefile'.
  Looking for an implicit rule for 'Makefile'.
  No implicit rule found for 'Makefile'.
  Finished prerequisites of target file 'Makefile'.
 No need to remake target 'Makefile'.
Updating goal targets....
Considering target file 'local-install'.
 File 'local-install' does not exist.
  Considering target file 'kind-load-image'.
   File 'kind-load-image' does not exist.
    Considering target file 'kind-setup'.
     File 'kind-setup' does not exist.
     Considering target file '/Users/simonbeck/repos/appcat/appcat-service-s3/.kind/kind-kubeconfig-v1.24.0'.
      File '/Users/simonbeck/repos/appcat/appcat-service-s3/.kind/kind-kubeconfig-v1.24.0' does not exist.
      Considering target file '/Users/simonbeck/go/bin/kind'.
       File '/Users/simonbeck/go/bin/kind' does not exist.
       Finished prerequisites of target file '/Users/simonbeck/go/bin/kind'.
      Must remake target '/Users/simonbeck/go/bin/kind'.
go install sigs.k8s.io/kind@latest
Putting child 0x7fe2a8f08f80 (/Users/simonbeck/go/bin/kind) PID 19734 on the chain.
Live child 0x7fe2a8f08f80 (/Users/simonbeck/go/bin/kind) PID 19734 
Reaping winning child 0x7fe2a8f08f80 PID 19734 
Removing child 0x7fe2a8f08f80 PID 19734 from chain.
      Successfully remade target file '/Users/simonbeck/go/bin/kind'.
      Finished prerequisites of target file '/Users/simonbeck/repos/appcat/appcat-service-s3/.kind/kind-kubeconfig-v1.24.0'.
     Must remake target '/Users/simonbeck/repos/appcat/appcat-service-s3/.kind/kind-kubeconfig-v1.24.0'.
/Users/simonbeck/go/bin/kind create cluster \
        --name provider-cloudscale-v1.24.0 \
        --image docker.io/kindest/node:v1.24.0 \
        --config kind/config.yaml
Putting child 0x7fe2a900c100 (/Users/simonbeck/repos/appcat/appcat-service-s3/.kind/kind-kubeconfig-v1.24.0) PID 19749 on the chain.
Live child 0x7fe2a900c100 (/Users/simonbeck/repos/appcat/appcat-service-s3/.kind/kind-kubeconfig-v1.24.0) PID 19749 
/bin/bash: /Users/simonbeck/go/bin/kind: No such file or directory
Reaping losing child 0x7fe2a900c100 PID 19749 
gmake: *** [kind/kind.mk:36: /Users/simonbeck/repos/appcat/appcat-service-s3/.kind/kind-kubeconfig-v1.24.0] Error 127
Removing child 0x7fe2a900c100 PID 19749 from chain.
```

Comment (Kidswiss) [reply kind/kind.mk:6]: But maybe don't invest too much time here. Rather implement the `$GOBIN` as discussed below. Maybe that already fixes the issues.

**Comment by ccremer** [comment]
TIL: This is how to create directories in make if it doesn't exist:
```
file: | dir
  touch $@

dir:
  mkdir -p $@
```
(note the pipe)
https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html

File: kind/kind.mk (hunk 1 above)
Comment (ccremer) [reply kind/kind.mk:7]: I've done the GOBIN part. Please try again.

**Comment by Kidswiss** [comment]
Looks like the hardcoding of `GOOS` and `GOARCH` in the `kind-load-image` step break things.

1. If `GOBIN` is set it doesn't seem to allow cross compilation: `go: cannot install cross-compiled binaries when GOBIN is set` (maybe that's also the reason it didn't work before)
2. It wants to cross compile...

Adding:
```
$(kind_bin): export GOOS = $(shell go env GOOS)
$(kind_bin): export GOARCH = $(shell go env GOARCH)
```

Seems to fix things.

**Comment by ccremer** [comment]
> Looks like the hardcoding of `GOOS` and `GOARCH` in the `kind-load-image` step break things.

Darn, forgot about that. I've re-added the env vars

**Review by Kidswiss — APPROVED** [review]

**Pull request merged**